plugins/inputs/cisco_telemetry_mdt_netconf/cisco_telemetry_mdt_netconf.go

```diff
@@ -0,0 +1,930 @@
+	// Store x-paths and tags from GetOperations
+	if gExist {
+		for _, g := range c.Grs.Gets {
+			if g.SelectFilter != "" {
+				c.userXpaths[c.trimXpath(g.SelectFilter)] = nil
+				for _, k := range g.Tags {
+					c.userTags[c.trimXpath(k)] = nil
+				}
+			}
+		}
+	}
+}
+
+//reportError abstracts the report of error messages based on their root cause
+func (c *CiscoTelemetryNETCONF) reportError(
+	ctx context.Context, err error, processName string) {
+	logMessage := "NETCONF %s stopped"
+	errorMessage := "failed to %s: %s"
+
+	if ctx.Err() != nil && ctx.Err() == context.Canceled {
+		// Acknowledge errors due to context cancellation
+		c.Log.Errorf(
+			logMessage, processName,
+		)
+		c.Log.Errorf("Context cancelled. Adding error to accumulator: ", err)
+		c.acc.AddError(fmt.Errorf(
+			errorMessage, processName, err,
+		))
+	} else if ctx.Err() != nil {
+		// Acknowledge other context error
+		c.Log.Errorf("Context error: Adding error to accumulator: ", err)
+		c.acc.AddError(fmt.Errorf(
+			errorMessage, processName, ctx.Err().Error(),
+		))
+
+	} else {
+		// Report base error
+		c.Log.Errorf("Adding error to accumulator: ", err)
+		c.acc.AddError(fmt.Errorf(
+			errorMessage, processName, err,
+		))
+	}
+}
+
+// cleanup should be called before the termination of the plugin
+func (s *setting) cleanup() error {
+	// Close open session and client before stopping the plugin
+	var err error
+	if s.Mutex != nil {
+		s.Mutex.Lock()
+		if s.Session != nil {
+			err = s.Session.Close()
+			// Override error if it is due to receive interruption, and hence
+			// invalid message framing
+			if err == netconf.ErrFraming {
+				err = nil
+			}
+		}
+		s.Mutex.Unlock()
+	}
+	if s.Client != nil {
+		if errorClient := s.Client.Close(); err == nil && errorClient != nil {
+			err = errorClient
+		}
+	}
+	return err
+}
+
+// Start the Cisco NETCONF service
+func (c *CiscoTelemetryNETCONF) Start(acc telegraf.Accumulator) error {
+	// Plugin context
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+
+	c.acc = acc
+
+	c.preparePaths()
+
+	// Create services for dial-in subscriptions and/or get requests
+	c.waitgroup.Add(1)
+	defer c.waitgroup.Done()
+	go func() {
+		if c.Dsrs != nil && len(c.Dsrs.Subscriptions) > 0 {
+			c.Log.Debugf(
+				"Starting Cisco NETCONF dial-in subscription service on %s",
+				c.ServerAddress)
+			c.waitgroup.Add(1)
+			go c.Dsrs.createService(ctx, c)
+		}
+		if c.Grs != nil && len(c.Grs.Gets) > 0 {
+			c.Log.Debugf(
+				"Starting Cisco NETCONF get service on %s",
+				c.ServerAddress,
+			)
+			c.waitgroup.Add(1)
+			go c.Grs.createService(ctx, c)
+		}
+		if c.Dsrs != nil && len(c.Dsrs.Notifications) > 0 {
+			c.Log.Debugf(
+				"Starting Cisco NETCONF notification subscription service on %s",
+				c.ServerAddress)
+			// Create the Dsrs service only if it was not already created
+			if len(c.Dsrs.Subscriptions) == 0 {
+				c.waitgroup.Add(1)
+				go c.Dsrs.createService(ctx, c)
+			}
+		}
+	}()
```

Comment:
Why do all those function call require an own go-routine? You are waiting for the service to be created anyway, so why not just run them in order without parallel execution. I don't think the time-gain is worth the effort, is it?